File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.java
class Solution {
    public int gcd(int a,int b){
        while(b!=0){
            int temp=b;
            b=a%b;
            a=temp;
        }
        return a;
    }
    public long gcdSum(int[] nums) {
        ArrayList<Integer> pre=new ArrayList<>();
        int mx=0;

        for(int x:nums){
            mx=Math.max(mx,x);
            pre.add(gcd(mx,x));
        }

        Collections.sort(pre);

        int i=0;
        int j=pre.size()-1;
        long ans=0;

        while(i<j){
            ans+=gcd(pre.get(i),pre.get(j));
            i++;
            j--;
        }

        return ans;
    }
}

File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.py
class Solution:
    def gcd(self,a, b):
        while b!=0:
            temp=b
            b=a%b
            a=temp
        return a
    
    def gcdSum(self, nums: list[int]) -> int:
        
        pre=[]
        mx=-inf
        for x in nums:
            mx=max(mx,x)
            pre.append(self.gcd(mx,x))

        pre.sort()

        i=0
        j=len(pre)-1
        ans=0

        while i<j:
            ans+=self.gcd(pre[i],pre[j])
            i+=1
            j-=1

        return ans


